packages/token/src/erc20/extensions/erc4626/erc4626.cairo

```diff
@@ -0,0 +1,603 @@
+    use crate::erc20::interface::{IERC20, IERC20Metadata};
+    use crate::erc20::interface::{IERC20Dispatcher, IERC20DispatcherTrait};
+    use openzeppelin_utils::math::Rounding;
+    use openzeppelin_utils::math;
+    use starknet::ContractAddress;
+    use starknet::storage::{StoragePointerReadAccess, StoragePointerWriteAccess};
+
+    // The default values are only used when the DefaultConfig
+    // is in scope in the implementing contract.
+    pub const DEFAULT_UNDERLYING_DECIMALS: u8 = 18;
+    pub const DEFAULT_DECIMALS_OFFSET: u8 = 0;
+
+    #[storage]
+    pub struct Storage {
+        ERC4626_asset: ContractAddress
+    }
+
+    #[event]
+    #[derive(Drop, PartialEq, starknet::Event)]
+    pub enum Event {
+        Deposit: Deposit,
+        Withdraw: Withdraw,
+    }
+
+    /// Emitted when `sender` exchanges `assets` for `shares` and transfers those
+    /// `shares` to `owner`.
+    #[derive(Drop, PartialEq, starknet::Event)]
+    pub struct Deposit {
+        #[key]
+        pub sender: ContractAddress,
+        #[key]
+        pub owner: ContractAddress,
+        pub assets: u256,
+        pub shares: u256
+    }
+
+    /// Emitted when `sender` exchanges `shares`, owned by `owner`, for `assets` and transfers
+    /// those `assets` to `receiver`.
+    #[derive(Drop, PartialEq, starknet::Event)]
+    pub struct Withdraw {
+        #[key]
+        pub sender: ContractAddress,
+        #[key]
+        pub receiver: ContractAddress,
+        #[key]
+        pub owner: ContractAddress,
+        pub assets: u256,
+        pub shares: u256
+    }
+
+    pub mod Errors {
+        pub const EXCEEDED_MAX_DEPOSIT: felt252 = 'ERC4626: exceeds max deposit';
+        pub const EXCEEDED_MAX_MINT: felt252 = 'ERC4626: exceeds max mint';
+        pub const EXCEEDED_MAX_WITHDRAW: felt252 = 'ERC4626: exceeds max withdraw';
+        pub const EXCEEDED_MAX_REDEEM: felt252 = 'ERC4626: exceeds max redeem';
+        pub const TOKEN_TRANSFER_FAILED: felt252 = 'ERC4626: token transfer failed';
+        pub const INVALID_ASSET_ADDRESS: felt252 = 'ERC4626: asset address set to 0';
+        pub const DECIMALS_OVERFLOW: felt252 = 'ERC4626: decimals overflow';
+    }
+
+    /// Constants expected to be defined at the contract level which configure virtual
+    /// assets and shares.
+    ///
+    /// `UNDERLYING_DECIMALS` should match the underlying asset's decimals. The default
+    /// value is `18`.
+    ///
+    /// `DECIMALS_OFFSET` corresponds to the representational offset between `UNDERLYING_DECIMALS`
+    /// and the vault decimals. The greater the offset, the more expensive it is for attackers to
+    /// execute an inflation attack.
+    ///
+    /// Requirements:
+    ///
+    /// - `UNDERLYING_DECIMALS`+ `DECIMALS_OFFSET` cannot exceed 255 (max u8).
+    ///
+    pub trait ImmutableConfig {
+        const UNDERLYING_DECIMALS: u8;
+        const DECIMALS_OFFSET: u8;
+
+        fn validate() {
```

Comment:
The reason for the having the user set `UNDERLYING_DECIMALS` was to avoid requiring recoverability from a call to `decimals`. If we'd prefer leveraging the forthcoming try/catch functionality, we can just automate the process like the [solidity impl](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/extensions/ERC4626.sol#L79-L97) and remove the user input for underlying